src/System.Management.Automation/engine/Utils.cs

```diff
@@ -1460,18 +1460,57 @@ public static void SetTestHook(string property, object value)
         }
     }
 
+    internal class HistoryStack<T>
+    {
+        private BoundedStack<T> _boundedUndoStack;
+        private BoundedStack<T> _boundedRedoStack;
+
+        internal HistoryStack(uint capacity)
+        {
+            _boundedUndoStack = new BoundedStack<T>(capacity);
+            _boundedRedoStack = new BoundedStack<T>(capacity);
+        }
+
+        internal void Push(T item)
+        {
+            _boundedUndoStack.Push(item);
+            if (RedoCount >= 0)
+            {
+                _boundedRedoStack.Clear();
+            }
+        }
+
+        internal T Undo(T currentItem)
+        {
+            var previousItem = _boundedUndoStack.Pop();
+            _boundedRedoStack.Push(currentItem);
+            return previousItem;
+        }
+
+        internal T Redo(T currentItem)
+        {
+            var nextItem = _boundedRedoStack.Pop();
+            _boundedUndoStack.Push(currentItem);
+            return nextItem;
+        }
+
+        internal int UndoCount => _boundedUndoStack.Count;
+
+        internal int RedoCount => _boundedRedoStack.Count;
+    }
+
     /// <summary>
-    /// An bounded stack based on a linked list.
+    /// A bounded stack based on a linked list.
     /// </summary>
     internal class BoundedStack<T> : LinkedList<T>
     {
-        private readonly int _capacity;
+        private readonly uint _capacity;
```

Comment:
I wonder - how do this related to the PR?

Reply:
As I was writing the XUnit unit tests for the 2 classes, I was thinking about the case of someone passing in a negative number, hence why I made it unsigned to get that check for free.